src/wh_server_keystore.c

```diff
@@ -14,37 +14,40 @@
 #include "wolfhsm/wh_packet.h"
 #include "wolfhsm/wh_error.h"
 
-int hsmGetUniqueId(whServerContext* server)
+int hsmGetUniqueId(whServerContext* server, whNvmId* outId)
 {
     int i;
     int ret = 0;
     whNvmId id;
+    /* apply client_id and type which should be set by caller on outId */
+    whNvmId buildId = ((*outId | (server->comm->client_id << 8)) & (~WOLFHSM_KEYID_MASK));
     whNvmId nvmId = 0;
     whNvmId keyCount;
     /* try every index until we find a unique one, don't worry about capacity */
-    for (id = 1; id < WOLFHSM_KEYID_MASK; id++) {
+    for (id = 1; id < WOLFHSM_KEYID_MASK + 1; id++) {
+        buildId = ((buildId & ~WOLFHSM_KEYID_MASK) | id);
         /* check against cache keys */
         for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
-            if ((id | WOLFHSM_KEYID_CRYPTO) == server->cache[i].meta->id)
+            if (buildId == server->cache[i].meta->id)
                 break;
         }
         /* try again if match */
         if (i < WOLFHSM_NUM_RAMKEYS)
             continue;
         /* if keyId exists */
         ret = wh_Nvm_List(server->nvm, WOLFHSM_NVM_ACCESS_ANY,
```

Comment:
Recommend to simply use wh_Nvm_GetMetadata() != WH_ERROR_NOTFOUND since you are only looking for an exact match.